File: Elixir/benchmark/lib/Benchmark/perfect_number.ex
# vim: ft=elixir sw=2 ts=2
#
# perfect_number.ex
# calculate perfect numbers
#

defmodule Benchmark.PerfectNumber do
  use Pipe

  @doc """
  Predicate checks whether a given number is perfect
  """
  def is_perfect?(n), do: is_perfect?(n, 1, 0)
  defp is_perfect?(n, i, sum) do
    if i < n do
      if rem(n, i) == 0 do
        is_perfect?(n, i+1, sum+i)
      else
        is_perfect?(n, i+1, sum)
      end
    else
      sum == n
    end
  end

  @doc """
  Returns a list of perfect numbers for given upper limit
  """
  def perfect_numbers(limit) do
    for x <- 1..limit, is_perfect?(x), do: x
  end

  @doc """
  Returns a list of perfect numbers for given upper limit.
  Calculations are performed asynchronous with Task.async
  """
  def perfect_numbers_async(limit) do
    kl = pipe_with fn(ll, f) -> Enum.map(ll, f) end,
      Enum.to_list(1..limit) |> spawn_task |> Task.await
    filter_true(kl)

    #tt = Enum.map(1..limit, fn(x) -> Task.async(fn -> {is_perfect?(x), x} end) end)
    #Keyword.get_values Enum.map(tt, fn(x) -> Task.await(x) end), true
  end

  defp spawn_task(x), do: Task.async(fn -> {is_perfect?(x), x} end)
  defp filter_true(kl), do: Keyword.get_values(kl, true)


end

File: Python/benchmark/benchmark/perfect_number.py
# vim: set ft=python sw=4 ts=4:

# perfect_number.py
# Calculate perfect numbers in Python (V3)

from typing import List
from multiprocessing import Pool

NUM_PROCESSES = 8
CHUNK_SIZE = 1

# predicate to check for perfect numbers


def is_perfect(n: int) -> bool:
    sum = 0
    for i in range(1, n):
        if (n % i) == 0:
            sum += i

    return sum == n


# generates a list of perfect numbers for given
# upper bound
def perfect_numbers(limit: int) -> List[int]:
    pn = []
    for i in range(1, limit):
        if is_perfect(i):
            pn.append(i)

    return pn

# coroutine generator for perfect numbers


def perfect_numbers_coroutine():
    counter = 1
    while True:
        if is_perfect(counter):
            yield counter
        counter += 1


def perfect_numbers_async(count: int) -> List[int]:
    pn = []
    gen = perfect_numbers_coroutine()
    # pn.append(next(gen))
    for _ in range(count):
        # pn.append(gen.send(None))
        pn.append(next(gen))

    gen.close()
    return pn


def perfect_numbers_mp(limit: int) -> List[int]:
    with Pool(processes=NUM_PROCESSES) as pool:
        pn_idx = pool.map(is_perfect, range(1, limit), chunksize=CHUNK_SIZE)
    pn = []
    for i, p in enumerate(pn_idx):
        if p:
            pn.append(i + 1)
    return pn
